[lldb-dap] Refactoring lldb-dap port listening mode to allow multiple connections.

ashgti opened the pull request:
This adjusts the lldb-dap listening mode to accept multiple clients.
Each client initializes a new instance of DAP and an associated `lldb::SBDebugger` instance.

The listening mode is configured with the `--connection` option and supports listening on a port or a unix socket on supported platforms.

vogelsgesang commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -116,6 +118,8 @@ enum LaunchMethod { Launch, Attach, AttachForSuspendedLaunch };
 /// Page size used for reporting addtional frames in the 'stackTrace' request.
 constexpr int StackPageSize = 20;
 
+void RegisterRequestCallbacks(DAP &dap);
```
seems to be never implemented and never called?

ashgti replied:
Removed.

labath commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1196,6 +1200,39 @@ def terminate(self):
+            env=adaptor_env,
+        )
+
+        if port or unix_socket:
+            # Wait for the server to startup.
+            time.sleep(0.1)
```
And this is the second race I was talking about.

ashgti replied:
Should be fixed with the new approach.

labath commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1154,34 +1154,38 @@ class DebugAdaptorServer(DebugCommunication):
-                self, s.makefile("r"), s.makefile("w"), init_commands
+                self, s.makefile("rb"), s.makefile("wb"), init_commands, log_file
+            )
+        elif unix_socket:
+            s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+            s.connect(unix_socket)
+            DebugCommunication.__init__(
+                self, s.makefile("rb"), s.makefile("wb"), init_commands, log_file
+            )
```
What's the advantage in supporting both kinds of sockets. Could we just do with one?

ashgti replied:
VS Code has helpers for running over stdout/stderr (https://code.visualstudio.com/api/references/vscode-api#DebugAdapterExecutable), a unix socket / named pipe (https://code.visualstudio.com/api/references/vscode-api#DebugAdapterNamedPipeServer) or tcp host:port (https://code.visualstudio.com/api/references/vscode-api#DebugAdapterServer).

Since `lldb-dap` had support for a port mode already, I didn't want to remove it. I added support for both protocols since it gives flexibility on how `lldb-dap` is used.

labath replied:
Supporting unix sockets *only* would not be good, as then this wouldn't work on windows. However, TCP/IP works everywhere, so it *might* be sufficient to only support those? I don't think its my place to tell you you shouldn't implement that, but I also don't want you to think that's a prerequisite for this feature (making sure it possible to add the functionality later is nice, but I think we could forego the actual implementation until someone actually needs it).

ashgti replied:
This should work with both tcp/unix sockets by parsing the `--connection` flag.

ashgti commented on lldb/packages/Python/lldbsuite/test/lldbtest.py:
```diff
@@ -250,6 +251,13 @@ def which(program):
 
+def pickrandomport():
+    """Returns a random open port."""
+    with socket.socket() as sock:
+        sock.bind(("", 0))
+        return sock.getsockname()[1]
```
I have updated this to not pass in the port, but instead read the port that was picked by `lldb-dap` when its picked.

Instead now you pass a `--connection` to `lldb-dap` and it will print the listeners its created for the connection like:

```
$ lldb-dap --connection tcp://localhost:0
Listening for: tcp://localhost:51282, tcp://localhost:51283
^C
$ lldb-dap --connection unix:///tmp/abc321
Listening for: unix:///tmp/abc321
``` 

The tests now read the port from stdout.

labath replied:
Cool. So I guess this means the function is unused now?

ashgti replied:
Reverted changes to this file.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -137,42 +141,232 @@ lldb::SBValueList *GetTopLevelScope(DAP &dap, int64_t variablesReference) {
+      *dap.log << error_message << std::endl;
+    output_callback_stderr(error_message);
+  }
+  dap.err = lldb::SBFile(new_stderr_fd.get(), "w", false);
```
Refactored this and made it crash return an error if the redirect fails.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -137,42 +141,232 @@ lldb::SBValueList *GetTopLevelScope(DAP &dap, int64_t variablesReference) {
+      *dap.log << error_message << std::endl;
+    output_callback_stderr(error_message);
+  }
+  dap.out = lldb::SBFile(new_stdout_fd.get(), "w", false);
```
Refactored this and made it crash return an error if the redirect fails.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -137,42 +141,232 @@ lldb::SBValueList *GetTopLevelScope(DAP &dap, int64_t variablesReference) {
+///
+/// \return
+///     A fd pointing to the original stdout.
+void SetupRedirection(DAP &dap, int stdoutfd = -1, int stderrfd = -1) {
```
Each `lldb::SBDebugger` instance has its own file for stdout/stderr.

When running in server mode we don't actually redirect stdout/stderr to the DAP connection directly.

When running lldb commands the output is directed to the `lldb::SBDebugger::GetOutputFileHandle` and `lldb::SBDebugger::GetErrorFileHandle`.

the SBDebugger defaults to stdout/stderr, but we can change those since we have one debugger per client connection.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -137,42 +141,232 @@ lldb::SBValueList *GetTopLevelScope(DAP &dap, int64_t variablesReference) {
+///
+/// \return
+///     A fd pointing to the original stdout.
+void SetupRedirection(DAP &dap, int stdoutfd = -1, int stderrfd = -1) {
```
Done.

ashgti commented on lldb/tools/lldb-dap/Options.td:
```diff
@@ -22,8 +22,17 @@ def port: S<"port">,
 
+def unix_socket: S<"unix-socket">,
+  MetaVarName<"<path>">,
+  HelpText<"Communicate with the lldb-dap tool over the unix socket or named pipe.">;
```
I refactored this to have a single `--connection` parameter.

The connection takes a description of the connection to listen on and supports the format `tcp://hostname:port` or `unix:///path`.

ashgti commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -140,13 +137,16 @@ struct DAP {
   std::thread event_thread;
   std::thread progress_event_thread;
-  std::unique_ptr<std::ofstream> log;
+  std::shared_ptr<std::ofstream> log;
```
Changed to a pointer instead.

ashgti commented on lldb/test/API/tools/lldb-dap/server/main.c:
```diff
@@ -0,0 +1,6 @@
+int main(int argc, char const *argv[]) {
+  printf("hello world!\n"); // breakpoint 1
+  return 0;
+}
```
Done.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5028,36 +5218,52 @@ int main(int argc, char *argv[]) {
+                            default_repl_mode, portno);
+  }
+
+  if (!unix_socket_path.empty()) {
```
I adjusted this to take a single `--connection` argument that we can use to create a listener for instead of two arguments.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -137,42 +141,232 @@ lldb::SBValueList *GetTopLevelScope(DAP &dap, int64_t variablesReference) {
+                     const std::vector<std::string> &pre_init_commands,
+                     std::shared_ptr<std::ofstream> log,
+                     ReplMode default_repl_mode, int portno) {
+  auto listening = getSocketFD(portno);
```
I refactored this and made some helpers for creating sockets and listening. LMKWYT

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -137,42 +141,232 @@ lldb::SBValueList *GetTopLevelScope(DAP &dap, int64_t variablesReference) {
+  struct sockaddr_in serv_addr;
+  bzero(&serv_addr, sizeof(serv_addr));
+  serv_addr.sin_family = AF_INET;
+  // serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
```
This was here before my change, but I refactored this and made some new helpers to support different kinds of sockets.

ashgti commented on lldb/tools/lldb-dap/Options.td:
```diff
@@ -22,8 +22,17 @@ def port: S<"port">,
+def unix_socket: S<"unix-socket">,
+  MetaVarName<"<path>">,
+  HelpText<"Communicate with the lldb-dap tool over the unix socket or named pipe.">;
+def: Separate<["-"], "u">,
```
Renamed this to `--connection` that handles different listening mode descriptions.

ashgti commented on lldb/tools/lldb-dap/Options.td:
```diff
@@ -22,8 +22,17 @@ def port: S<"port">,
+def: Separate<["-"], "u">,
+  Alias<unix_socket>,
+  MetaVarName<"<path>">,
+  HelpText<"Alias for --unix_socket">;
```
Renamed this to `--connection` that handles different listening mode descriptions.

ashgti commented on lldb/tools/lldb-dap/Options.td:
```diff
@@ -22,8 +22,17 @@ def port: S<"port">,
+def: Separate<["-"], "u">,
+  Alias<unix_socket>,
+  MetaVarName<"<path>">,
+  HelpText<"Alias for --unix_socket">;
```
Renamed this to `--connection` that handles different listening mode descriptions.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -32,35 +34,44 @@ using namespace lldb_dap;
 
+llvm::Error DAP::ConfigureIO(int out_fd, int err_fd) {
+  llvm::Expected<int> new_stdout_fd =
+      RedirectFd(out_fd, [this](llvm::StringRef data) {
+        SendOutput(OutputType::Stdout, data);
+      });
```
AFAICT, nothing ensures these callbacks don't run after the DAP object was terminated. This wasn't a big deal when the process was serving only one connection (although making the object stack-allocated in #116272 is kinda problematic), but I think this needs to be sorted out (ideally as a separate patch) for the multi-connection version.

It also doesn't look like the pipes created here get cleaned up anywhere.

walter-erquinigo replied:
Yeah, this is maybe the trickiest part of all. The redirection was introduced to prevent LLDB proper from printing messages to stderr and unexpectedly causing the DAP client to abort. And indeed, there are many parts of LLDB that print directly to stderr upon unusual situations (dwarf parsing is one of them IIRC). The ideal solution is to create an actual SB API that can handle these unusual messages and print to stderr is simply the default behavior. You could also do that.
If you don't do such refactor, something that you should consider when using multiple connections simultaneously, is that it might be tricky to know which of those stderr messages correspond to which DAP client. You might decide to multicast them, which is better than sending the output to the wrong client. Silencing the messages altogether might not be a good idea.
In any case, you can think of the redirection as something that needs to be set up only once per process.

ashgti replied:
Adjusted the RedirectFD to clean up once the DAP object is not used anymore. This includes making a best effort to use `dup2` to revert the FD redirects.


As for stderr handling, in server mode it no longer setups a redirect on stderr to the current connection. When I allocate the SBDebugger I also configure the OutputFile and ErrorFile handles to pipes that forward to that specific connection.

This does mean that if we are running in a server mode and some part of lldb (or llvm/clang/etc.) print directly to stderr then it wouldn't be associated with the debug session, but I think that would be an issue for other users of the SBDebugger API. 

I know that Xcode also has a helper `lldb-rpc-server` that is a similar configuration to the lldb-dap server mode. When a debug session starts in Xcode it starts an `lldb-rpc-server` that hosts all the SBDebugger instances. So a single version of Xcode will have 1 instance of `lldb-rpc-server` running at a time and that instance is used to run each debug session for each Xcode window. But my overall point with this is that if there are parts of the stack that are not respecting the SBDebugger's OutputFile and ErrorFile it would also show up in Xcode.

labath replied:
> once the DAP object is not used anymore

I don't think you can really guarantee that while there are detached forwarding threads floating around. I think we need to shut those down first. The tricky part is doing that in cross-platform way, but since we're already all-in on lldbHost, I think we could/should use lldb's Pipe::ReadWithTimeout to implement polled reads (which check for a "should terminate flag" in between). Not the most elegant solution, but I it's probably the best we can do with what we have right now. One of these days I'm going to implement a MainLoop<->Pipe integration that can multiplex reads (and other things), but it's not very easy due to very different Pipe APIs on windows vs. unix.

I also believe this parts should be a separate patch, as there's a lot of subtleness involved.

ashgti replied:
I've refactored the IO handling in #122783 to ensure the DAP instance + SBDebugger has its own dedicated file handles for the DAP session.

labath commented on lldb/tools/lldb-dap/IOStream.h:
```diff
@@ -52,11 +51,11 @@ struct StreamDescriptor {
-  bool read_full(std::ofstream *log, size_t length, std::string &text);
+  bool read_full(llvm::raw_ostream *log, size_t length, std::string &text);
 
-  bool read_line(std::ofstream *log, std::string &line);
+  bool read_line(llvm::raw_ostream *log, std::string &line);
 
-  bool read_expected(std::ofstream *log, llvm::StringRef expected);
+  bool read_expected(llvm::raw_ostream *log, llvm::StringRef expected);
```
I'm not sure why is this being changed, but could that be a separate patch as well?

ashgti replied:
Reverted this change.

labath replied:
(Note I'm not saying this isn't a good change. I'd be happy to review something like this as a separate patch. I just don't want to mix this in with the rest of the changes, as there's a lot of things going on here already.)

labath commented on lldb/tools/lldb-dap/Socket.h:
```diff
@@ -0,0 +1,182 @@
+//===-- Socket.h ------------------------------------------------*- C++ -*-===//
```
I'm not quite comfortable with this essentially reimplementing large parts of `source/Host/common/*Socket.cpp`.  @walter-erquinigo do you have any objections to using the implementation from the lldb `host` library. I see lldb-dap already depends on `lldbHost` on windows (and NetBSD?) -- this would make that dependency unconditional.

ashgti replied:
Sounds good, I can re-work my patch to use `lldb/Hosts/*`

ashgti replied:
I reworked this now to use the `lldb/Host/Socket.h` and associated types.

labath replied:
Cool. Thanks.

labath reviewed (commented):
The patch is still a bit larger than I'd like. I've tried to highlight a few things that could be split off into separate PRs, but I think the biggest question now (not really for you, but for lldb-dap maintainers) is whether this could be using the socket abstractions already present in lldb (instead of building your own). If the answer to that is "yes", then switching to those should probably be a separate patch as well.

labath commented:
> > The patch is still a bit larger than I'd like. I've tried to highlight a few things that could be split off into separate PRs, but I think the biggest question now (not really for you, but for lldb-dap maintainers) is whether this could be using the socket abstractions already present in lldb (instead of building your own). If the answer to that is "yes", then switching to those should probably be a separate patch as well.
> 
> I can split this into smaller patches.

Thanks.

> 
> As far as the socket abstraction, at the moment lldb-dap doesn't use much (if anything) from lldb outside of the API folder. I saw the `lldb/Host/Socket.h` files but they also are built on top of the `IOObject` and interact with the `MainLoop` which would mean taking on a few new dependencies from the Host folder. 

I am aware of that, but I don't think that situation is sustainable in the long run. When lldb-dap was first added, I believe it only supported talking over stdio, so it had like ~zero amount of socket code, and so this kind of purist approach did not hurt (much). However, its getting quite complex now, and I think copying/reimplementing every abstraction it needs is going to come back and bite us.

FWIW, the MainLoop class is self-contained, and might actually be useful if you wanted to support things like listening on IPv4 and v6 sockets simultaneously. And IOObject is just a abstract base, so it also doesn't shouldn't pull in much. We're also making sure (it took a lot of effort to reach that point) that the Host library as a whole does not depend on the rest of lldb.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5028,36 +5218,52 @@ int main(int argc, char *argv[]) {
+                            default_repl_mode, unix_socket_path);
+  }
+
+  DAP dap = DAP(program_path.str(), log, default_repl_mode, pre_init_commands);
```
Done, made this a uniform callback for both stdin/stdout and connections.

ashgti commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1196,6 +1202,62 @@ def terminate(self):
+            adaptor_env["LLDBDAP_LOG"] = log_file
+
+        if os.uname().sysname == "Darwin":
+            adaptor_env["NSUnbufferedIO"] = "YES"
```
Removed, having a flush seems to fix fix this.

ashgti commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1196,6 +1202,62 @@ def terminate(self):
+                print("Reading stdout for the listening connection")
+                os.set_blocking(proc.stdout.fileno(), False)
+                stdout_key = sel.register(proc.stdout, selectors.EVENT_READ)
+                rdy_fds = sel.select(timeout=10.0)
```
I refactored this to not use a select and instead adjusted the script to not buffer when using a connection. This lets us read a single line at a time without reaching EOF first.

ashgti commented on lldb/test/API/tools/lldb-dap/server/TestDAP_server.py:
```diff
@@ -0,0 +1,67 @@
+        """
+        self.do_test_server(connection="tcp://localhost:0")
+
+    def test_server_unix_socket(self):
```
Done.

ashgti commented on lldb/test/API/tools/lldb-dap/server/TestDAP_server.py:
```diff
@@ -0,0 +1,67 @@
+        """
+        Test launching a binary with a lldb-dap in server mode on a unix socket.
+        """
+        dir = tempfile.gettempdir()
```
Done.

labath commented:
> I can split this into smaller patches.

Please, do that. Multithreading and servers and stuff are very tricky to get right, and it'd be a lot easier to review this if it was in smaller chunks. I think there's a lot of prepwork so that the code is prepared to handle multiple connections. That stuff can land without actually supporting multiple connections.

ashgti commented on lldb/source/Host/common/TCPSocket.cpp:
```diff
@@ -118,6 +124,15 @@ std::string TCPSocket::GetRemoteConnectionURI() const {
   return "";
 }
 
+std::string TCPSocket::GetListeningConnectionURI() const {
```
Split this into its own PR https://github.com/llvm/llvm-project/pull/118330

ashgti committed:
[lldb-dap] Refactoring lldb-dap port listening mode to allow multiple connections.

This adjusts the lldb-dap listening mode to accept multiple clients.
Each client initializes a new instance of DAP and an associated `lldb::SBDebugger` instance.

The listening mode is configured with the `--connection` option and supports listening on a port or a unix socket on supported platforms.

ashgti commented on lldb/source/Host/common/TCPSocket.cpp:
```diff
@@ -180,8 +195,9 @@ Status TCPSocket::Listen(llvm::StringRef name, int backlog) {
-      host_port->hostname.c_str(), nullptr, AF_UNSPEC, SOCK_STREAM, IPPROTO_TCP);
+  std::vector<SocketAddress> addresses =
+      SocketAddress::GetAddressInfo(host_port->hostname.c_str(), nullptr,
+                                    AF_UNSPEC, SOCK_STREAM, IPPROTO_TCP);
```
Reverted

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5028,48 +5021,128 @@ int main(int argc, char *argv[]) {
+      if (log)
+        *log << "configureIO failed: " << llvm::toStringWithoutConsuming(Err)
+             << "\n";
+      std::cerr << "failed to configureIO: " << llvm::toString(std::move(Err))
```
I am using `llvm::errs()` instead for reporting errors when running in server mode.

labath replied:
That's good but beware `os << error` does not "consume" (set the ["checked" bit](https://llvm.org/docs/ProgrammersManual.html#recoverable-errors)), on the llvm::Error object (which means it will trigger an assert in debug mode). You need to consume it somehow, e.g., via toString, or `logAllUnhandledErrors` (the latter works best when the entire "main" function is wrapped in something returning an error object

ashgti commented:
I've updated this after splitting parts of this off into multiple smaller patches (#118330, #118565, #122783)

labath reviewed (commented):
I think this looks pretty good now. Just a couple of questions from my side.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5058,72 +5053,148 @@ int main(int argc, char *argv[]) {
+            StreamDescriptor::from_socket(client->GetNativeSocket(), false),
+            StreamDescriptor::from_socket(client->GetNativeSocket(), false),
+            /*=redirectOut*/ nullptr, /*=redirectErr*/ nullptr);
+      }).detach();
```
Moved this to a ThreadPool so we do not need to detach the threads.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5058,72 +5053,148 @@ int main(int argc, char *argv[]) {
+            name,
+            StreamDescriptor::from_socket(client->GetNativeSocket(), false),
+            StreamDescriptor::from_socket(client->GetNativeSocket(), false),
+            /*=redirectOut*/ nullptr, /*=redirectErr*/ nullptr);
```
Done

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -4952,6 +4930,28 @@ static int DuplicateFileDescriptor(int fd) {
 
+static llvm::Expected<std::pair<Socket::SocketProtocol, std::string>>
+parseConnection(llvm::StringRef conn) {
+  if (conn.contains("://")) {
```
Updated this to validate the connection and used the URI Parser to parse the connection.

ashgti commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -1167,21 +1168,61 @@ def __init__(
+                    out.removeprefix(expected_prefix).rstrip("\r\n").split(",", 1)[0]
+                )
+
+            if connection.startswith("unix-connect://"):  # unix-connect:///path
```
Done.

labath commented:
Overall, I like the StreamDescriptor changes, but I fear this patch is starting to lose focus again. Could you separate the refactor into a separate PR?

ashgti commented:
> Overall, I like the StreamDescriptor changes, but I fear this patch is starting to lose focus again. Could you separate the refactor into a separate PR?

Reverted those changes, I kept running into clang (Xcode 16.0 release of clang) crashes compiling lldb-dap for some reason with the existing StreamDescriptor and the `OnAccept` callback with the `MainLoop` object. This version is working for me now and not crashing the compiler.

labath commented on lldb/test/API/tools/lldb-dap/server/Makefile:
```diff
@@ -0,0 +1,3 @@
+C_SOURCES := main.c
+
+include Makefile.rules
```
I think this stop sign indicates a missing newline.

ashgti replied:
Done.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -58,12 +57,13 @@ const char DEV_NULL[] = "/dev/null";
       output(std::move(output)), broadcaster("lldb-dap"),
-      exception_breakpoints(), focus_tid(LLDB_INVALID_THREAD_ID),
-      stop_at_entry(false), is_attach(false),
+      exception_breakpoints(), pre_init_commands(pre_init_commands),
```
```suggestion
      exception_breakpoints(), pre_init_commands(std::move(pre_init_commands)),
```

ashgti replied:
Applied

labath commented on lldb/tools/lldb-dap/DAPForward.h:
```diff
@@ -11,6 +11,8 @@
 
 // IWYU pragma: begin_exports
 
+#include "lldb/lldb-forward.h"
```
Probably revert this too?

ashgti replied:
Reverted

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5058,72 +5053,144 @@ int main(int argc, char *argv[]) {
 
-  DAP dap = DAP(program_path.str(), log.get(), default_repl_mode,
-                std::move(input), std::move(output));
+    error = listener->Listen(name, /* backlog */ 5);
```
```suggestion
    error = listener->Listen(name, /*backlog=*/5);
```

ashgti replied:
Applied

labath reviewed (commented):
I think this looks good, modulo the question about (detached) threads. I'd be nice to have an lldb-dap expert take a look as well.

ashgti committed:
Applying review comments.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5058,72 +5053,144 @@ int main(int argc, char *argv[]) {
+    // address.
+    llvm::outs().flush();
+
+    llvm::DefaultThreadPool pool(llvm::optimal_concurrency());
```
I misread the `optimal_concurrency` and thought it didn't limit the number of threads created. Switched this to the default `hardware_concurrency` which doesn't limit the number of threads.

I had previously tried using `SIGHUP` to try to have the server shutdown all the active sessions but I ran into some issues with the current DAP terminate call and the flow of things between the debugger. I think that is something we can add in a follow up patch. That way the server has some way of performing a 'clean' shutdown if desired.

ashgti replied:
I again misunderstood how the limit flag was applied to the thread pool.

I moved away from the thread pool and changed this to track the `DAP *` objects directly. With that change I was also able to have a clean shutdown when the server is interrupted by registering an interrupt handler with `llvm::sys::SetInterruptFunction` to trigger the shutdown.

ashgti committed:
Ensuring the server mode tracks active DAP sessions and adding an interrupt handler to disconnect active sessions.

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5058,72 +5018,187 @@ int main(int argc, char *argv[]) {
+      }
+    }
+
+    return client_failed ? EXIT_FAILURE : EXIT_SUCCESS;
```
There's a race here between this thread exiting (and destroying the `active_dap_sessions` list) and the now-disconnected threads trying to remove themselves from it.

I think that could be fixed by waiting until this list becomes empty (ideally using the aforementioned [std::notify_all_at_thread_exit](https://en.cppreference.com/w/cpp/thread/notify_all_at_thread_exit) to delay cond_var notification until the thread is well into shutting itself down.

ashgti replied:
Done.

labath reviewed (commented):
Apart from the comment, this looks good to me. @vogelsgesang, @walter-erquinigo, any thoughts from you?

ashgti committed:
Using a std::condition_variable to notify the main thread after all DAP sessions have disconnected to coordinate shutdowns.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5058,72 +5018,187 @@ int main(int argc, char *argv[]) {
+      g_loop.AddPendingCallback(
+          [](lldb_private::MainLoopBase &loop) { loop.RequestTermination(); });
+    });
+    std::mutex active_dap_sessions_mutext;
```
Done.

ashgti committed:
Refactoring to TestDAP_server tests to not create an extra client connection on startup of the binary.

ashgti commented:
@walter-erquinigo Any updates on allowing multiple connections in server mode?

ashgti commented:
@JDevlieghere Any thoughts on the server mode?

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -4952,6 +4895,29 @@ static int DuplicateFileDescriptor(int fd) {
+      conn.str().c_str());
+}
+
 int main(int argc, char *argv[]) {
```
the newly added logic seems correct, but would it be possible to extract out some of code into other functions? I'm afraid main is starting to get too big to be easily readable

ashgti replied:
I moved the main connection handling loop to its own function, we can break things up further as well if you'd like

walter-erquinigo reviewed (approved):
sorry for the late review, this LGTM
I suggested refactoring main a bit. If you can do it, it would be great, but all good if you don't. But next time we add more stuff to main, it would be good simplify that function a bit

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -58,12 +58,13 @@ const char DEV_NULL[] = "/dev/null";
+DAP::DAP(std::string name, llvm::StringRef path, std::ofstream *log,
+         StreamDescriptor input, StreamDescriptor output, ReplMode repl_mode,
+         std::vector<std::string> pre_init_commands)
+    : name(name), debug_adaptor_path(path), log(log), input(std::move(input)),
```
```suggestion
    : name(std::move(name)), debug_adaptor_path(path), log(log), input(std::move(input)),
```

ashgti replied:
Done.

JDevlieghere reviewed (approved):
LGTM

ashgti committed:
Moving connection handling to its own helper to simplify main and applying reviewer suggestions.

ashgti committed:
Rewording an error message.

ashgti commented:
https://github.com/llvm/llvm-project/pull/128278 Should mark the test as skipped on windows to unblock the CI.

kazutakahirata commented:
@ashgti I've fixed the warnings from this PR with 434d86ac929a37b2e992329d4e81fa0fbcacd2a9.  Thanks!
